specs/bellatrix/fork-choice.md

```diff
@@ -78,6 +79,100 @@ As per EIP-3675, before a post-transition block is finalized, `notify_forkchoice
+) -> boolean:
+    justified_state = store.checkpoint_states[store.justified_checkpoint]
+    head_block = store.blocks[head_root]
+    parent_root = head_block.parent_root
+    parent_block = store.blocks[parent_root]
+    current_slot = get_current_slot(store)
+
+    # Single slot re-org.
+    parent_slot_ok = parent_block.slot + 1 == head_block.slot
+    time_into_slot = (store.time - store.genesis_time) % SECONDS_PER_SLOT
+    current_time_ok = (head_block.slot == current_slot or
+                       (head_block.slot + 1 == current_slot and
+                        time_into_slot < SECONDS_PER_SLOT // INTERVALS_PER_SLOT))
+    single_slot_reorg = parent_slot_ok and current_time_ok
+
+    # Shuffling stable.
+    shuffling_stable = (head_block.slot + 1) % SLOTS_PER_EPOCH != 0
+
+    # FFG information of the new head_block will be competitive with the current head.
+    # FIXME: needs more justification/finalization information in the store
+    # ffg_competitive =
+    #    store.realized_justifications[parent_root] ==
+    #        store.realized_justifications[head_root] and
+    #    store.realized_finalizations[parent_root] ==
+    #        store.realized_finalization[head_root]
+    ffg_competitive = True
+
+    # Ensure that total participation meets a minimum threshold.
+    # The participation multiplier is set to the number of slots for which the parent_block block has
+    # been known.
+    participation_multiplier = current_slot - parent_block.slot
+    parent_weight = get_latest_attesting_balance(store, parent_root)
+    participation_threshold = calculate_committee_fraction(justified_state, PARTICIPATION_THRESHOLD)
+    participation_ok = parent_weight >= participation_multiplier * participation_threshold
+
+    # Check the head weight only if the attestations from the head slot have already been applied.
+    # Implementations may want to do this in different ways, e.g. by advancing
+    # `store.time` early, or by counting queued attestations during the head block's slot.
+    if current_slot > head_block.slot:
+        assert store.proposer_boost_root == Root()
+        head_weight = get_latest_attesting_balance(store, head_root)
+        reorg_threshold = calculate_committee_fraction(justified_state, REORG_WEIGHT_THRESHOLD)
+        head_weak = head_weight < reorg_threshold
+    else:
+        head_weak = True
+
+    # Only re-org the head_block block if it arrived later than the attestation deadline.
+    head_late = store.block_timeliness.get(head_root) is False
+
+    # Only suppress the fork choice update if we are confident that we will propose the next block.
+    parent_state_advanced = store.block_states[parent_root]
+    process_slots(parent_state_advanced, head_block.slot + 1)
+    proposer_index = get_beacon_proposer_index(parent_state_advanced)
+    proposing_reorg_slot = validator_is_connected(proposer_index)
+
+    return all([single_slot_reorg, shuffling_stable, ffg_competitive, participation_ok,
```

Comment:
In the current Prysm proposal, we are missing `shuffling_stable`, `ffg_competitive`, `participation_ok` and part of single slot reorg conditions. I'll circle back with @potuz

Reply:
We do have the latter two, but we don't have the shuffling safety check.

Reply:
Verygood👌🌹🙏

Reply:
Verygood👌🌹🙏